src/month_delta.rs

```diff
@@ -0,0 +1,105 @@
+use chrono::{Datelike, NaiveDate};
+use polars::prelude::*;
+
+// Function to get the date of the last day of the current month for a given date.
+fn get_last_month_date(date: NaiveDate) -> NaiveDate {
+    if date.month() == 12 {
+        // If it is December, move to the next year and set the month to January.
+        NaiveDate::from_ymd_opt(date.year() + 1, 1, 1)
+            .unwrap()
+            .pred_opt()
+            .unwrap()
+    } else {
+        date.with_day(1)
+            .unwrap()
+            .with_month(date.month() + 1)
+            .unwrap()
+            .pred_opt()
+            .unwrap()
+    }
+}
+
+// Function that checks if both dates fall on the last days of their respective months.
+fn get_last_day_bool(start_date: NaiveDate, end_date: NaiveDate) -> bool {
+    let end_date_end = get_last_month_date(end_date);
+    let start_date_end = get_last_month_date(start_date);
+    {
+        // End date is the last day of its month
+        end_date.day() == end_date_end.day() &&
+            // Start date is the last day of its month
+            start_date.day() == start_date_end.day() &&
+            end_date.day() != start_date.day() &&
+            start_date.month() != end_date.month()
+    }
+}
+
+// Function to calculate the span of months between two dates as an integer.
+// This function specifically checks if the span between the two dates covers whole months,
+// and under certain conditions, adjusts the count by 1 or -1 to reflect partial months.
+fn get_month_span_int(start_date: NaiveDate, end_date: NaiveDate) -> i32 {
+    // Check if the actual number of days difference matches assuming both 
+    // dates start on the first
+    let actual_days_diff = end_date.signed_duration_since(start_date).num_days();
+    let expected_days_diff = {
+        let start_dt = start_date.with_day(1).unwrap(); // start date at the beginning of the month
+        let end_dt = end_date.with_day(1).unwrap(); // end date at the beginning of a month
+        end_dt.signed_duration_since(start_dt).num_days() // expected day difference as full months
+    };
+
+    // Calculates if the date difference spans entire months
+    // If do then add additional month to the calculation
+    if actual_days_diff == expected_days_diff
+        && end_date.month() != start_date.month()
+        && end_date.day() != start_date.day()
+    {
+        1
+    } else if expected_days_diff.abs() > actual_days_diff.abs() {
+        // If the expected difference (in absolute terms) is greater than the actual difference,
+        // it indicates a partial month span, and we return -1 to adjust the month span downwards.
+        -1
+    } else {
+        // If none of the conditions were met
+        0
+    }
+}
+
+pub(crate) fn impl_month_delta(start_dates: &Series, end_dates: &Series) -> PolarsResult<Series> {
```

Comment:
if you have a start_date and end_date, is it possible to do something like (pseudocode):

```
n = 0
while start_date < end_date:
    start_date += 1_month
    n += 1
return n
```
?

might simplify things a bit?

Reply:
thanks! Polars has `add_us` / `add_ms` / `add_ns` function though that can already accept months, I wonder if it's possible to use those? As well as Polars' own `Duration`?

Reply:
@MarcoGorelli Just wanted to confirm, you mean rewrite everything to use Polars' `Duration` or this specific function?

Reply:
as in, I think this whole function could be done a lot more simply by using polars' `add_ms` / `add_us` / `add_ns` and Duration, with an algorithm like
```
n = 0
while start_date < end_date:
    start_date += 1_month
    n += 1
return n
```